src/Groups/matrices/linear_centralizer.jl

```diff
@@ -0,0 +1,426 @@
+
+
+########################################################################
+#
+# Orders
+#
+########################################################################
+
+function _GL_order(n::Int, q::fmpz)
+   res = q^div(n*(n-1),2)
+   for i in 1:n res *= (q^i-1) end
+   return res
+end
+
+_GL_order(n::Int, F::Ring) = _GL_order(n, order(F))
+
+function _SL_order(n::Int, q::fmpz)
+   res = q^div(n*(n-1),2)
+   for i in 2:n res *= (q^i-1) end
+   return res
+end
+
+_SL_order(n::Int, F::Ring) = _SL_order(n, order(F))
+
+########################################################################
+#
+# Centralizer in GL
+#
+########################################################################
+
+# Compute a generating set for GL(n,q) and SL(n,q)
+# generators for GL(n,q) and SL(n,q) are described in: 
+# Taylor, D. E., Pairs of Generators for Matrix Groups. I, The Cayley Bulletin, 3 (1987)
+
+# returns as matrices
+# returns a generating set for GL(n,F) of at most two elements
+function _gens_for_GL(n::Int, F::FinField)
+   n !=1 || return [matrix(F,1,1,[primitive_element(F)])]
+   h1 = identity_matrix(F,n)
+   h2 = zero_matrix(F,n,n)
+   if order(F)==2
+      h1[1,2] = 1
+   else
+      h1[1,1] = primitive_element(F)
+      h2[1,1] = -1
+   end
+   h2[1,n] = 1
+   for i in 1:n-1 h2[i+1,i] = -1 end
+   return h1,h2
+end
+
+# returns as matrices
+# does the same as above with the following changes:
+# the field F is replaced by F = F[x]/(f);
+# every entry y is replaced by a diagonal join of D copies of phi(y),
+# where phi is the ring automorphism sending a fixed root of f into the companion matrix of f.
+# (hence, the final output are still matrices with base ring F)
+# ASSUMPTION: deg(f) > 1
+function _gens_for_GL_matrix(f::PolyElem, n::Int, F::FinField; D::Int=1)
+   C = companion_matrix(f)
+   CP = evaluate(_centralizer(f),C)            # matrix of maximal order in the centralizer of the companion matrix
+
+   if n==1 return [diagonal_join([CP for i in 1:D])] end
+   h1 = identity_matrix(F,n*degree(f)*D)
+   insert_block!(h1,diagonal_join([CP for i in 1:D]),1,1)
+   h2 = zero_matrix(F,n*degree(f)*D,n*degree(f)*D)
+   for i in 1:(n-1)*degree(f)*D h2[i+degree(f)*D,i]=-1 end
+   for i in 1:degree(f)*D
+      h2[i,i]=-1
+      h2[i,i+(n-1)*degree(f)*D]=1
+   end
+   return h1,h2
+end
+
+# generators for centralizer of unipotent elements are described in:
+# Giovanni De Franceschi, Centralizers and conjugacy classes in finite classical groups, arXiv:2008.12651
+
+# V = vector of integers of the dimensions of Jordan blocks
+# return the generators for the centralizers of the diagonal join of unipotent Jordan blocks of dimensions V
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+function _centr_unipotent(F::FinField, V::AbstractVector{Int}; isSL=false)
+   n = sum(V)
+   _lambda = gen(F)  # yes, gen(F) is correct; we don't need a primitive root in this case
```

Comment:
Just for my understanding: So what is `gen(F)`, then?!? Perhaps an element such that its powers form a basis of `F` over the prime field / base field?

I hink you need at least *something* for _lambda, otherwise you could use `one(F)` or even `zero(F)` here; might be a good idea to state in a code comment what it is you do need, not just what you do *not* need... :-)

Reply:
`gen(a::FqFiniteField)` :  Return the generator of the finite field.
So, `gen(F)` generates `F` as a ring, namely every element of `F` can be written as sum of powers of `gen(F)`. So, `one(F)` is not good whenever `degree(F)` is greater than 1. The difference with `Hecke.primitive_element(F)` is that `gen(F)` does not need to be a generator for the multiplicative group of `F`. I expect that computing `gen(F)` is much simpler than computing `Hecke.primitive_element(F)`, so in this case I used `gen(F)`, because that's all I need.